src/magic/model/MagicAbility.java

```diff
@@ -633,6 +633,12 @@ protected void addAbilityImpl(final MagicAbilityStore card, final Matcher arg) {
             ));
         }
     },
+    Ascend("ascend",0) {
+        @Override
+        protected void addAbilityImpl(final MagicAbilityStore card, final Matcher arg) {
+            card.add(OtherEntersBattlefieldTrigger.Ascend);
```

Comment:
using an ETB trigger doesn't work for Ascend as you may control 10 permanents through an effect that gains control of an opponent's creature.

Reply:
Pretty sure I missed something.

So I should add a new trigger type `AscendTrigger`, then I think I'll need to add `MagicTriggerType.Ascend` for it, and execute this trigger type in `PlayCardFromStackAction` and `ChangeControlAction`. Is this right?

Reply:
I'm not sure change control and ETB are the only two possibility there may be others as well. What we want to check is that a particular condition is true, and not that a certain action occurred. As describe in the rules, Ascend on permanent is a static ability, so it should be a static ability at the game layer with the `accept` method checking for "you control 10 or more permanents and no blessing" and the `modGame` method performing the action to give the blessing. It will be similar to the MagicStatic created by MagicStatic.StateTrigger.

Reply:
I see `MagicStatic.StateTrigger` uses `PutStateTriggerOnStackAction`, then search for that in card scripts and found how it's implemented in [Immortal Coil](https://github.com/magarena/magarena/blob/master/release/Magarena/scripts/Immortal_Coil.groovy). So I'll implement ascend like that as well.